sites/main-site/src/content/blog/2024/seqera-containers-part-1.mdx

```diff
@@ -0,0 +1,453 @@
+https://conda.anaconda.org/conda-forge/linux-64/libgcc-14.1.0-h77fa898_1.conda#002ef4463dd1e2b44a94a4ace468f5d2
+# .. truncated ..
+```
+
+### Trust and transparency
+
+Reproducibility is one thing, but it's also important that the images we use can be trusted and are secure.
+Users and developers must be able to verify that the contents of the container match the `environment.yml` for the module.
+
+The first point of trust is the `community.wave.seqera.io` base URI.
+This is only used for Seqera Containers and only `wave.seqera.io` is able to push images.
+It is also more locked-down than regular Wave, only allowing Conda builds and not custom Dockerfiles or augmented images.
+
+Second, the image URI includes a tag that has a hash of the input files used to request the image.
+We can use this to retrieve build details, which we will add to the nf-core website module page:
+
+-   Build time, architecture and platform (Docker / Singularity)
+-   `Dockerfile` / Singularity recipe and Conda `environment.yml`
+-   Full build logs
+-   Conda lock files with full package URLs and md5sums for entire dependency resolution
+-   For Docker images: Trivy security scan and software bill of materials (SBOM)
+
+:::info{title="Example Wave build details page" collapse}
+
+We haven't built this yet, but here's what the Wave build details page looks like, which has the same information:
+
+<Image
+    class="d-block ms-3 mb-3 float-end rounded shadow"
+    src={wave_logs}
+    alt="Screenshot of a Wave container build details page."
+/>
+
+Wave build details page for the MultiQC v1.24.1 image: `community.wave.seqera.io/library/multiqc:1.24.1--789bc3917c8666da`
+
+The final part is the hash - from this we can infer the Wave Build ID and retrieve the build details: [https://wave.seqera.io/view/builds/789bc3917c8666da_1](https://wave.seqera.io/view/builds/789bc3917c8666da_1)
+:::
+
+The new conda lock files also allow enhanced reproducibility for Conda users and more stable CI tests - we'll come back to these in a future blog post / bytesize.
+
+Finally, the image creation will happen automatically on GitHub Actions when an `environment.yml` file is edited.
+This means that the entire flow from conda file through to final containers will be entirely transparent, as the request to Wave itself and its response will be available in the GitHub Actions logs.
+
+### No lock-in
+
+Using Seqera Containers does not mean that nf-core will be locked into using Seqera tooling.
+The suggested implementation has two components:
+
+1.  Automated on-demand generation of container images using Wave
+2.  Long term hosting images on Seqera Containers
+
+The tooling for image generation is [open-source](https://github.com/seqeralabs/wave).
+Seqera is hosting this API and build service for the community for free, but there's nothing to stop us from hosting the same API ourselves in the future.
+As such, there is no lock-in on this API or any tooling we will write around it.
```

Comment:
Makes me wonder if we should have a roll-back plan if something goes wrong with the mass replacement system (unless this is going to be a gradual change a la nf-test?)

Reply:
Nah, should be fine. The mass replacement is a one-off. I was thinking about this the other day. I think the best is the following:

* I generate a list of every conda env file to be requested
* I warn the Seqera team that it's coming, so they can bump up the back-end cluster size ahead of time.
* I use the Wave CLI to request every conda env file 4 times (docker, singularity, amd and arm)

I can do this whenever. Then when we come to migrate the modules, they'll each request a new image but it will already exist, so the API will just return the built cache immediately and cause no stress on the system. Going forward it'll be incremental so shouldn't be an issue.

Should be fun trying to request this many images in one go 😬 Curious what the final count is..